packages/test/amplify_auth_integration_test/lib/src/totp_utils.dart

```diff
@@ -0,0 +1,73 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'package:amplify_auth_cognito_dart/amplify_auth_cognito_dart.dart';
+import 'package:amplify_flutter/amplify_flutter.dart';
+import 'package:collection/collection.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:otp/otp.dart';
+
+/// The device name given to the default TOTP authenticator.
+const friendlyDeviceName = 'friendlyDeviceName';
+
+/// The interval between TOTP codes, i.e. how long before they expire.
+const Duration _totpInterval = Duration(seconds: 30);
+
+/// Calculates the timestamp of the next TOTP code, based off the previous one, if any.
+Future<int> get _nextTotpTime async {
+  final now = DateTime.now();
+  if (OTP.lastUsedTime == 0) {
+    addTearDown(() {
+      OTP.lastUsedTime = 0;
+      OTP.lastUsedCounter = 0;
+    });
+    return now.millisecondsSinceEpoch;
+  }
+  // Cognito allows a +/- 1 interval range for codes and codes cannot be reused.
```

Comment:
I did not know that TOTP codes cannot be reused within the window. It is part of [the spec](https://www.rfc-editor.org/rfc/rfc6238) though. TIL.

> Note that a prover may send the same OTP inside a given time-step
   window multiple times to a verifier.  The verifier MUST NOT accept
   the second attempt of the OTP after the successful validation has
   been issued for the first OTP, which ensures one-time only use of an
   OTP.

Reply:
Took me forever to figure this out. Ended up having to read the Cognito source 😅